### PR TITLE
fix(a380x/fms): window constraints, deletion, alt formatting

### LINKED ISSUE #9297: Unable to edit or add Altitude Constraints to waypoints
### Aircraft Version

Development

### Build info

```json
{
    "built": "2024-10-31T15:54:41+00:00",
    "ref": "refs/tags/v0.12.0",
    "sha": "2fe89113291f2478347dde0dc3228b55dc924f74",
    "actor": "FoxtrotSierra6829",
    "event_name": "manual",
    "pretty_release_name": "stable/v0.12.0",
    "version": "a380x-v0.12.0-rel.2fe8911"
}
```

### Describe the bug

Unable to edit constraints that are added automatically on STAR's. In a lot of cases you might be told to decent to a flight level above what is published on the charts because of Transition Layers etc. For example FL80 instead of FL70.
It allows you to type them in but no way of confirming and inserting it. 
Screenshot_1
Screenshot_2
Screenshot_3

### Expected behavior

You should be able to edit any constraint as ATC can give you instructions that differ from the published charts.

### Steps to reproduce

1. Insert any STAR that has Altitude Constraints
2. Attempt to edit an Altitude Constraint
3. Unable to progress any further with the edit

### References (optional)

_No response_

### Additional info (optional)

Sorry if it's already been flagged up, couldn't see it on the issues

### Discord Username (optional)

#15.11

### PR DESCRIPTION
**Pull request by tracernz**
Fixes #9297

## Summary of Changes

- Fix entry of transition altitude on the PERF CLB page (copied code from PERF TO page).
- Fix usage of transition altitude; descent constraints should use transition level.
- Fixed window constraint handling and disabling of constraint edition on window constraints.
- Fixed deletion of constraints.
- Noted some other issues while reading through the code for a future PR (issues opened for both).

Best to review one commit at a time.

## Additional context

Discord username (if different from GitHub):

## Testing instructions

Try to reproduce #9297; the constraints should now be editable (but once you edit a window constraint you can not get a window back again).
Try to deletion both and altitude and a speed constraint with the button on the constraint page; they should delete and be removed from the page immediately.

## How to download the PR for QA

Every new commit to this PR will cause new A32NX and A380X artifacts to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, find and click on the **PR Build** tab
1. Click on either **flybywire-aircraft-a320-neo**, **flybywire-aircraft-a380-842 (4K)** or **flybywire-aircraft-a380-842 (8K)** download link at the bottom of the page

### CODE REVIEW

File: fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFplnVertRev.tsx [hunk 1]
```diff
@@ -195,32 +195,40 @@ export class MfdFmsFplnVertRev extends FmsPage<MfdFmsFplnVertRevProps> {
...
-          this.altWindowUnitLeading.set('');
-          this.altWindowUnitTrailing.set('');
-          this.altWindowUnitValue.set('');
-        }
+
+        // FIXME check format when only the higher altitude is above TA/TL
+        const alt1IsFl = transAlt !== null && ac.altitude1 > transAlt;
+        const alt2IsFl = transAlt !== null && ac.altitude2 > transAlt;
+        this.altWindowUnitLeading.set(alt1IsFl && !alt2IsFl ? 'FT' : '');
+        this.altWindowUnitTrailing.set(alt1IsFl ? 'FL' : 'FT');
```
Comment (Benjozork) [inline fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFplnVertRev.tsx:226]: **question:** why would the trailing unit ever be `FL`? Is that how it actually works IRL?

File: fbw-a32nx/src/systems/fmgc/src/flightplanning/legs/FlightPlanLeg.ts [hunk 2]
```diff
@@ -303,17 +303,26 @@ export class FlightPlanLeg implements ReadonlyFlightPlanLeg {
     return this;
   }
 
-  clearConstraints() {
-    this.definition.verticalAngle = undefined;
+  clearAltitudeConstraints() {
     this.definition.altitudeDescriptor = undefined;
```
Comment (Benjozork) [inline fbw-a32nx/src/systems/fmgc/src/flightplanning/legs/FlightPlanLeg.ts:307]: **thought:** 😬 we should really use `.editLegDefinition` to accomplish this... for later I guess. this currently breaks sync

File: fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFplnVertRev.tsx [hunk 3]
```diff
@@ -195,32 +195,40 @@ export class MfdFmsFplnVertRev extends FmsPage<MfdFmsFplnVertRevProps> {
...
 
-      if (leg.altitudeConstraint?.altitude2 !== undefined) {
-        const ac = leg.altitudeConstraint;
-        this.altConstraintDisabled.set(true);
-        // ALT window
-        const transAlt = this.transitionAltitude.get();
-        if (ac.altitude1 && transAlt && ac.altitude1 > transAlt) {
-          // FL
-          this.altWindowUnitLeading.set('FL');
+      const ac = leg.altitudeConstraint;
```
[the remark this answers is not in the document]
Comment (tracernz) [reply fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFplnVertRev.tsx:209]: I don’t disagree, but I also didn’t change it. 😁

File: fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFplnVertRev.tsx (hunk 1 above)
Comment (tracernz) [reply fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFplnVertRev.tsx:226]: The second one is the highest altitude.

Comment (Benjozork) [reply fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFplnVertRev.tsx:226]: So.. it shows like, `10000 FT - 210 FL`?

**Review by flogross89 — APPROVED** [review]

**Comment by 2hwk** [comment]
Quality Assurance Tester Report

Discord Username: 2cas
Object of testing: #9459
Aircraft:  A380X
Tier of Testing: 1
Date: 15/11/2024 

Testing Process: 
- [x] Try to reproduce https://github.com/flybywiresim/aircraft/issues/9297; the constraints should now be editable (but once you edit a window constraint you can not get a window back again).
- [x] Try to deletion both and altitude and a speed constraint with the button on the constraint page; they should delete and be removed from the page immediately.
- [x] Deleted and recreated altitude and speed constraints throughout the flight plan
- [x] Switched FMS and attempted to modify constraints again

Testing Results: 
QA Passed

Negatives:       
- N/A

Conclusions: 
- LGTM

**Pull request merged**